### PR TITLE
Slicing improvements

### PR DESCRIPTION
**Pull request by msperber**
- In Python, allows using numpy-like syntax to select ranges and strides across multiple axes, e.g. ```my_expr[:,1:4,::2]``` will now work. Under the hood, this uses the strided_select operator.
- as a special case, when using this to select a range of batch elements, and this range starts with the first batch element, the operation is marked as in-placed. Ideally, this should also work when the range does not start at the first batch element, but that would require some adjusting some pointers so I'll leave it for a future PR.

### CODE REVIEW

**Comment by neubig** [comment]
Awesome, this will be very useful! Could you add a test for this?

Also maybe check this:
https://app.codacy.com/app/xunzhang/dynet/pullRequest?prid=1590987

**Comment by msperber** [comment]
Sure, I've added some tests.

File: tests/python/test.py [hunk 1]
```diff
@@ -459,6 +459,20 @@ def test_layer_norm(self):
         self.assertTrue(np.allclose(y.npvalue(), y_np_value))
 
 
+class TestSlicing(unittest.TestCase):
+
+    def test_layer_norm(self):
```
Comment (neubig) [inline tests/python/test.py]: The name of this function is wrong, but other than that this is great! I'm happy to merge once this is fixed.

**Comment by msperber** [comment]
Thanks for catching this, should be fixed now.

**Pull request merged**